ext/src/ruby_api/wasi_config.rs

```diff
@@ -0,0 +1,283 @@
+        let mut inner = rb_self.get()?.inner.borrow_mut();
+        inner.stdin = Some(ReadStream::Path(path));
+        Ok(rb_self)
+    }
+
+    /// @yard
+    /// Set stdin to the specified String.
+    /// @param content [String]
+    /// @def set_stdin_string(content)
+    /// @return [WasiConfig] +self+
+    pub fn set_stdin_string(rb_self: RbSelf, content: RString) -> Result<RbSelf, Error> {
+        let mut inner = rb_self.get()?.inner.borrow_mut();
+        inner.stdin = Some(ReadStream::String(content));
+        Ok(rb_self)
+    }
+
+    /// @yard
+    /// Inherit stdout from the current Ruby process.
+    /// @return [WasiConfig] +self+
+    pub fn inherit_stdout(rb_self: RbSelf) -> Result<RbSelf, Error> {
+        let mut inner = rb_self.get()?.inner.borrow_mut();
+        inner.stdout = Some(WriteStream::Inherit);
+        Ok(rb_self)
+    }
+
+    /// @yard
+    /// Set stdout to write to a file. Will truncate the file if it exists,
+    /// otherwise try to create it.
+    /// @param path [String] The path of the file to write to.
+    /// @def set_stdout_file(path)
+    /// @return [WasiConfig] +self+
+    pub fn set_stdout_file(rb_self: RbSelf, path: RString) -> Result<RbSelf, Error> {
+        let mut inner = rb_self.get()?.inner.borrow_mut();
+        inner.stdout = Some(WriteStream::Path(path));
+        Ok(rb_self)
+    }
+
+    /// @yard
+    /// Inherit stderr from the current Ruby process.
+    /// @return [WasiConfig] +self+
+    pub fn inherit_stderr(rb_self: RbSelf) -> Result<RbSelf, Error> {
+        let mut inner = rb_self.get()?.inner.borrow_mut();
+        inner.stderr = Some(WriteStream::Inherit);
+        Ok(rb_self)
+    }
+
+    /// @yard
+    /// Set stderr to write to a file. Will truncate the file if it exists,
+    /// otherwise try to create it.
+    /// @param path [String] The path of the file to write to.
+    /// @def set_stderr_file(path)
+    /// @return [WasiConfig] +self+
+    pub fn set_stderr_file(rb_self: RbSelf, path: RString) -> Result<RbSelf, Error> {
+        let mut inner = rb_self.get()?.inner.borrow_mut();
+        inner.stderr = Some(WriteStream::Path(path));
+        Ok(rb_self)
+    }
+
+    /// @yard
+    /// Set env to the specified +Hash+.
+    /// @param env [Hash<String, String>]
+    /// @def set_env(env)
+    /// @return [WasiConfig] +self+
+    pub fn set_env(rb_self: RbSelf, env: RHash) -> Result<RbSelf, Error> {
+        let mut inner = rb_self.get()?.inner.borrow_mut();
+        inner.env = Some(env);
+        Ok(rb_self)
+    }
+
+    /// @yard
+    /// Set the arguments (argv) to the specified +Array+.
+    /// @param args [Array<String>]
+    /// @def set_argv(args)
+    /// @return [WasiConfig] +self+
+    pub fn set_argv(rb_self: RbSelf, argv: RArray) -> Result<RbSelf, Error> {
+        let mut inner = rb_self.get()?.inner.borrow_mut();
+        inner.args = Some(argv);
+        Ok(rb_self)
+    }
```

Comment:
>What do you find ambiguous from inherit args and inherit env?

In Ruby, you can mutate the `ENV` and `ARGV`. [Changes to `ENV` are propagated to the process' env](https://github.com/ruby/ruby/blob/4491bb740a9506d76391ac44bb2fe6e483fec952/hash.c#L7261-L7265), but `ARGV` aren't. It's ambiguous (to me) which is the expected behaviour: should we use the Ruby-mutated ARGV, or should we use the original args (which segfaults with Rust' implementation, oops!).

By not providing the feature and letting integrators inject ENV & ARGV, we have a smaller API surface and remove the ambiguity.


-----


> Could we provide an api that simulates the inheriting process by ready from `ARGV` / `ENV` under the hood?

Yeah, totally, it'd be quite easy. I didn't see a lot of value in doing it though, as per the above.

Reply:
> Also should the explicit setters be named `args` and `envs`? My only reason for suggesting that is that they map directly to how that API is surface in `WasiCtxBuilder`.

I took example on the C WASI API which uses [`wasi_config_set_argv`](https://github.com/bytecodealliance/wasmtime/blob/75cd888e232aad18ac1c7bd0f40a6929ffba30c6/crates/c-api/src/wasi.rs#L106) and [`wasi_config_set_env`](https://github.com/bytecodealliance/wasmtime/blob/75cd888e232aad18ac1c7bd0f40a6929ffba30c6/crates/c-api/src/wasi.rs#L125).

We borrow pieces from the Rust and C APIs. Similar to Rust's `WasiCtxBuilder`, `WasiConfig` has a builder-style API. Similar to the C API's `wasi_config_t`, `WasiConfig` holds a configuration which is later applied.

Weakly-held opinion: `env` and `argv` is slightly better because it maps more closely to the Ruby constants, and there's a precedent in the C API anyways. Thoughts?

Reply:
> n Ruby, you can mutate the ENV and ARGV. Changes to ENV are propagated to the process' env, but ARGV aren't. It's ambiguous (to me) which is the expected behaviour: should we use the Ruby-mutated ARGV, or should we use the original args (which segfaults with Rust' implementation, oops!).

> By not providing the feature and letting integrators inject ENV & ARGV, we have a smaller API surface and remove the ambiguity.


I assumed this was referring to the `Rust` api or api as a whole. This makes sense to me. Keeping the `set_<argv|env>` makes sense.